Add Varanidae keyboard

TokageItLab opened the pull request:
<!--- Provide a general summary of your changes in the title above. -->

<!--- This template is entirely optional and can be removed, but is here to help both you and us. -->
<!--- Anything on lines wrapped in comments like these will not show up in the final text. -->

## Description

<!--- Describe your changes in detail here. -->

## Types of Changes

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply. -->
- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [x] Keyboard (addition or update)
- [x] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

zvecr commented on keyboards/varanidae/varanidae.h:
```diff
@@ -0,0 +1,42 @@
+
+#include "quantum.h"
+
+#define LAYOUT( \
```
Per PR checklist:

> LAYOUT macros are no longer accepted and should instead be moved to info.json

zvecr commented on keyboards/varanidae/rules.mk:
Needs to match `data/templates/keyboard/rules.mk`

```make
# This file intentionally left blank
```

zvecr commented on keyboards/varanidae/readme.md:
```diff
@@ -0,0 +1,18 @@
+    make varanidae:default
+    make varanidae:via
+
+To reset the keyboard into bootloader mode, press the reset switch on the Elite-C.
+
+See the [build environment setup](https://docs.qmk.fm/#/getting_started_build_tools) and the [make instructions](https://docs.qmk.fm/#/getting_started_make_guide) for more information. Brand new to QMK? Start with our [Complete Newbs Guide](https://docs.qmk.fm/#/newbs).
```
```suggestion
See the [build environment setup](https://docs.qmk.fm/#/getting_started_build_tools) and the [make instructions](https://docs.qmk.fm/#/getting_started_make_guide) for more information. Brand new to QMK? Start with our [Complete Newbs Guide](https://docs.qmk.fm/#/newbs).

## Bootloader

Enter the bootloader in 3 ways:

* **Bootmagic reset**: Hold down the key at (0,0) in the matrix (usually the top left key or Escape) and plug in the keyboard
* **Physical reset button**: Briefly press the reset switch on the Elite-C
* **Keycode in layout**: Press the key mapped to `QK_BOOT` if it is available
```

zvecr commented on keyboards/varanidae/keymaps/default/keymap.c:
```diff
@@ -0,0 +1,63 @@
+};
+
+#ifdef ENCODER_ENABLE
+bool encoder_update_kb(uint8_t index, bool clockwise) {
```
You cannot implement `_kb` callbacks at the keyboard level. These should be moved to `keyboards/varanidae/varanidae.c`

zvecr commented on keyboards/varanidae/info.json:
```diff
@@ -0,0 +1,152 @@
+        "device_version": "1.0.0"
+    },
+    "pin_compatible": "elite_c",
+    "processor": "atmega32u4",
+    "bootloader": "atmel-dfu",
```
```suggestion
"development_board": "elite_c",
```

zvecr commented on keyboards/varanidae/info.json:
```diff
@@ -0,0 +1,152 @@
+            { "pin_a": "D0", "pin_b": "D1", "resolution": 2 }
+        ]
+    },
+    "mouse_key": {
+        "enabled": true
+    },
```
```suggestion
```

zvecr commented on keyboards/varanidae/info.json:
```diff
@@ -0,0 +1,152 @@
+    "processor": "atmega32u4",
+    "bootloader": "atmel-dfu",
+    "features": {
+        "dip_switch": true,
```
```suggestion
        "bootmagic": true,
        "extrakey": true,
        "mousekey": true,
        "dip_switch": true,
```

zvecr commented on keyboards/varanidae/keymaps/via/keymap.c:
```diff
@@ -0,0 +1,87 @@
+        KC_PAST, KC_P4,  KC_P5,   KC_P6,   KC_END,    KC_LALT,        KC_A,  KC_S,  KC_D,  KC_F,  KC_G,  KC_H,  KC_J,  KC_K,    KC_L,   KC_SCLN, KC_QUOT,  KC_BSLS,
+        KC_PMNS, KC_P1,  KC_P2,   KC_P3,   KC_PGUP,   KC_LSFT, KC_BSLS,  KC_Z,  KC_X,  KC_C,  KC_V,  KC_B,  KC_N,  KC_M,  KC_COMM, KC_DOT, KC_SLSH, KC_GRV,   KC_RSFT,
+        KC_PPLS, KC_P0,  KC_PDOT, KC_PENT, KC_PGDN,   KC_LWIN, KC_LCTL, KC_PSCR, KC_INT5,                KC_SPC,                KC_INT4, KC_LEFT, KC_UP, KC_DOWN, KC_RIGHT
+    ),
+    [1] = LAYOUT(
+                 KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS,  KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS,
+        KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS,  KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS,
+        KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS,  KC_TRNS,   KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS,  KC_TRNS,
+        KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS,  KC_TRNS,     KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS,      KC_TRNS,
+        KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS,  KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS,    KC_TRNS,
+        KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS,  KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS,                KC_TRNS,                KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS
+    ),
+    [2] = LAYOUT(
+                 KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS,  KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS,
+        KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS,  KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS,
+        KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS,  KC_TRNS,   KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS,  KC_TRNS,
+        KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS,  KC_TRNS,     KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS,      KC_TRNS,
+        KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS,  KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS,    KC_TRNS,
+        KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS,  KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS,                KC_TRNS,                KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS
+    ),
+    [3] = LAYOUT(
+                 KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS,  KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS,
+        KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS,  KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS,
+        KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS,  KC_TRNS,   KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS,  KC_TRNS,
+        KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS,  KC_TRNS,     KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS,      KC_TRNS,
+        KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS,  KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS,    KC_TRNS,
+        KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS,  KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS,                KC_TRNS,                KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS
+    )
```
```suggestion
    )
```

zvecr commented on keyboards/varanidae/keymaps/via/keymap.c:
```diff
@@ -0,0 +1,87 @@
+    )
+};
+
+#ifdef ENCODER_ENABLE
```
Same issue as default keymap

waffle87 commented on keyboards/varanidae/readme.md:
```diff
@@ -0,0 +1,24 @@
+# Varanidae
+
+![Varanidae](https://imgur.com/a/GB1pG02)
```
```suggestion
![Varanidae](https://i.imgur.com/0wWfmE1h.png)
```

waffle87 commented on keyboards/varanidae/varanidae.c:
```diff
@@ -0,0 +1,52 @@
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "varanidae.h"
```
```suggestion
#include "quantum.h"
```

waffle87 commented on keyboards/varanidae/varanidae.h:
This file should be removed.

drashna commented on keyboards/varanidae/varanidae.c:
```diff
@@ -0,0 +1,52 @@
+
+#include "quantum.h"
+
+#ifdef ENCODER_ENABLE
+bool encoder_update_kb(uint8_t index, bool clockwise) {
+    if (!encoder_update_user(index, clockwise)) {
+        return false;
+    }
+    if (index == 0) {
+        if (clockwise) {
+            tap_code(KC_MS_WH_DOWN);
+        } else {
+            tap_code(KC_MS_WH_UP);
+        }
+    }
+    return true;
+}
+#endif
```
Isn't strictly needed now, as there is default behavior, unless you want to explicitly set it as a mousewheel by default.  Otherwise, it's volume control if extra keys are enabled, or scroll keys if mousekeys are enable but extrakeys are not.

TokageItLab replied:
Since the concept of this keyboard was to be developed for working with Blender, it is intended to function as a wheel scroll in most cases. Since we hold the pen tablet in our right hand, we needed the scroll wheel somewhere to change blender inspector value.

TokageItLab replied:
Adopted .encoder_map`.

drashna commented on keyboards/varanidae/varanidae.c:
```diff
@@ -0,0 +1,52 @@
+    }
+    switch (index) {
+        case 0: {
+            if (active) {
```
Is the dip switch an actual dip switch?  If so, then this is ... a bit odd for behavior.  If it's not an actual dip switch, then this may be benefit better from a custom matrix with D2 integrated into the normal switch matrix.

TokageItLab replied:
This is intended to be used as a middle click in most cases.

If the middle click is placed in the same matrix as the other keys, I could not be sure that it would always act in parallel when multiple keys are pressed, so I had to use a dip switch hack to prevent the circuit from competing with the other keys.

sigprof replied:
Placing the encoder push switch in the matrix would have worked perfectly fine — the diodes in the matrix ensure that any combination of keys pressed at the same time can be detected without ghosting issues, and holding a key for a long time is not a problem by itself.  Most boards which have encoders do things like that (often wiring the encoder press switch in parallel to the pads for an MX switch that can be alternatively installed in the same location).

Any key rollover issues that you may be thinking about don't actually exist, because the 6KRO restriction applies only to the key events that are actually sent to the host using the 6KRO report, and in case of `KC_MS_BTN3` those events are not sent in the keyboard report (the mouse report is completely separate).

The issues with using the DIP Switch feature for this purpose are:
1. Reassigning the function of the switch now needs some special code which is unlike any other board (in particular, you can't use tools like QMK Configurator or VIA for that).
2. DIP switches are not debounced, therefore you may encounter chatter issues when abusing that feature for a press switch.

Unfortunately, QMK currently does not offer an easy way to combine a “direct pin” key (what you effectively have for the encoder push switch now) with a key matrix.  The generic way to solve that problem is implementing a custom matrix, which is somewhat annoying (and you would need to copy most of the default matrix code).

However, one reasonably easy workaround for just a single direct pin key is to declare that pin as a matrix column for the `COL2ROW` matrix (if you had a `ROW2COL` matrix, you would need to declare that pin as a matrix row instead).  In a `COL2ROW` matrix column pins work as inputs, and connecting a matrix input to GND would result in reporting all key locations in the corresponding column as pressed; then you can specify one of these locations as a key in the layout, and ignore others (they would be mapped to `KC_NO` by the layout macro generated from the layout definition in `info.json`).  As an optimization, you may also use the `MATRIX_MASKED` feature to mask away those ignored locations (so that the QMK code would not need to look up the keycodes and process the `KC_NO` mappings for them).

The extra column workaround wastes some space in the keymap, which might be important for VIA usage (you may find that the same number of layers no longer fits into the EEPROM).  If that is important, and you actually have unused locations in other parts of the matrix, the only way (apart from a PCB redesign to put the encoder push switch into the key matrix) would be to implement a custom matrix which reports the state of the push switch using some otherwise unused matrix location.

sigprof replied:
Also the choice to use the `B0` pin on the Elite-C was somewhat unfortunate, because that pin is considered to be a design mistake of Elite-C (it occupies the place where Pro Micro had the RAW power pin, and therefore makes Elite-C not 100% compatible with Pro Micro, because some boards made for Pro Micro use the RAW pin).  Now this pin choice makes your PCB incompatible with most (maybe all) replacements for Elite-C with more capable MCUs (RP2040-based: Elite-Pi, Liatris, Splinky, Helios, …; STM32F4x1: STeMCell).  So if you are considering a “rev2” of the PCB, you really should think about moving the encoder switch into the matrix and then using the `D2` pin in the matrix instead of `B0`.

TokageItLab replied:
> Any key rollover issues that you may be thinking about don't actually exist, because the 6KRO restriction applies only to the key events that are actually sent to the host using the 6KRO report, and in case of KC_MS_BTN3 those events are not sent in the keyboard report (the mouse report is completely separate).

I did not know that about it. Thanks for the information.

> As an optimization, you may also use the MATRIX_MASKED feature to mask away those ignored locations (so that the QMK code would not need to look up the keycodes and process the KC_NO mappings for them).
The extra column workaround wastes some space in the keymap, which might be important for VIA usage (you may find that the same number of layers no longer fits into the EEPROM).

I tried to use MATRIX_MASKED but I gave up because I encountered an error that seemed to be over the limit of the EEPROM.

> So if you are considering a “rev2” of the PCB, you really should think about moving the encoder switch into the matrix and then using the D2 pin in the matrix instead of B0.

This keyboard already has 110 keys in the normal matrix. This is the upper limit of 11x10=110, and if we do not use the B0 switch, there is no space to put the encoder switch within the regular matrix. So it will be necessary to adopt some kind of anomalous matrix, but that is too hard for me.

The two problems mentioned by @sigprof is exist, but since there is no problem for the current use, so I would like to settle on this as the behavior for rev1.

sigprof replied:
> This keyboard already has 110 keys in the normal matrix. This is the upper limit of 11x10=110, and if we do not use the B0 pin, there is no space to put the encoder switch within the regular matrix.

Ah, I did not try to actually look whether there are any empty spaces in the matrix (this is one thing which is actually easier to do with the old-style layout macros — any locations containing `KC_NO` are immediately visible).  So your board actually is at the limits of Elite-C even when using the `B0` pin (and QMK core still does not support placing the rotary encoders inside the switch matrix, therefore the encoder requires two dedicated pins for the rotation signals).

> I tried to use MATRIX_MASKED but I gave up because I encountered an error that seemed to be over the limit of the EEPROM.

The error is not actually related to `MATRIX_MASKED`, but just to the fact that your matrix becomes 11×11 after adding a new column, so each layer needs 22 more bytes in the EEPROM, and there is not enough EEPROM to store that data.  The usual solution for that is limiting the dynamic keymap to 3 layers (by default 4 layers are used) — add this snippet to `info.json`:
```json
    "dynamic_keymap": {
        "layer_count": 3
    },
```
If I did not miss anything, the fixed QMK+VIA settings currenty use 41 bytes in the EEPROM, so for your original configuration (without encoder mapping) the remaining EEPROM size for VIA macros would be `1024 - 41 - 11*10*2*4` = 103 bytes, which is very close to the limit of 100 bytes, and in fact enabling the encoder mapping (as I suggested in the review) would put you over the limit anyway, so you would really need to remove one layer.  With 3 layers, extra column for the encoder push switch and encoder mapping you would have `1024 - 41 - (11*11 + 2)*2*3` = 245 bytes remaining for macros.

TokageItLab replied:
For encoder rotation, I cleaned up the encoder rotation by using `encoder map`. Thank you.

However, as for the encoder push, I could not find a suitable way to integrate it into the matrix (could not find a way to identify the independent circuit signals as an element of the matrix), so it will remain to use the dip switch hack.

TokageItLab replied:
Resolved! https://github.com/qmk/qmk_firmware/pull/21340#issuecomment-1605773482

TokageItLab commented:
The 6U key in layout was incorrectly located in `info.json` and I fixed it.

As for the two concerns by @drashna, I believe I gave fair reasons for both, but if there is a better way to implement this that could be resolved by adding some options to `info.json`, please let me know.

sigprof commented on keyboards/varanidae/keymaps/via/rules.mk:
```diff
@@ -0,0 +1 @@
+VIA_ENABLE = yes
```
To make the encoder rotation configurable in VIA, you need to enable the encoder map support:
```suggestion
VIA_ENABLE = yes
ENCODER_MAP_ENABLE = yes
```
(and also include the encoder map in `keymap.c`, otherwise the code won't compile).

Actually you can do the same changes in the `default` keymap too, so that the way to configure the encoder would be obvious to some user who decides to copy the `default` keymap before doing their modifications.

sigprof commented on keyboards/varanidae/keymaps/via/keymap.c:
```diff
@@ -0,0 +1,29 @@
+        KC_PMNS, KC_P1,  KC_P2,   KC_P3,   KC_PGUP,   KC_LSFT, KC_BSLS,  KC_Z,  KC_X,  KC_C,  KC_V,  KC_B,  KC_N,  KC_M,  KC_COMM, KC_DOT, KC_SLSH, KC_GRV,   KC_RSFT,
+        KC_PPLS, KC_P0,  KC_PDOT, KC_PENT, KC_PGDN,   KC_LWIN, KC_LCTL, KC_PSCR, KC_INT5,                KC_SPC,                KC_INT4, KC_LEFT, KC_UP, KC_DOWN, KC_RIGHT
+    )
+};
```
```suggestion
};

#if defined(ENCODER_MAP_ENABLE)
const uint16_t PROGMEM encoder_map[][NUM_ENCODERS][NUM_DIRECTIONS] = {
    [0] =  { ENCODER_CCW_CW(KC_MS_WH_UP, KC_MS_WH_DOWN) },
};
#endif
```

sigprof commented on keyboards/varanidae/info.json:
```diff
@@ -0,0 +1,151 @@
+        "pid": "0x0001",
+        "device_version": "1.0.0"
+    },
+    "pin_compatible": "elite_c",
```
This is not really correct, because the definition of `elite_c` compatibility in QMK does not include the `B0` pin of Elite-C, and your PCB uses that pin.  However, there is no way to explicitly set “none” here (only the `"promicro"` and `"elite_c"` values are allowed by the JSON schema), and `"development_board": "elite_c"` implicitly sets `"pin_compatible": "promicro"` for historical reasons.  So maybe `"pin_compatible": "elite_c"` would need to stay here for now (but the code won't compile anyway if a converter is selected, because the converters don't provide a mapping for the `B0` pin).

TokageItLab replied:
Using the B0 pin is currently unavoidable and must be left as is.

sigprof commented on keyboards/varanidae/info.json:
```diff
@@ -0,0 +1,151 @@
+                { "label":"K038", "matrix": [8, 3], "x": 18.25, "y": 1.25, "w": 1 },
+                { "label":"K039", "matrix": [7, 4], "x": 19.25, "y": 1.25, "w": 1 },
+
+                { "label":"K040", "matrix": [2, 0], "x": 0, "y": 2, "w": 1 },
+                { "label":"K041", "matrix": [2, 1], "x": 1, "y": 2, "w": 1 },
+                { "label":"K042", "matrix": [2, 2], "x": 2, "y": 2, "w": 1 },
+                { "label":"K043", "matrix": [2, 3], "x": 3, "y": 2, "w": 1 },
+                { "label":"K044", "matrix": [2, 4], "x": 4, "y": 2, "w": 1 },
+                { "label":"K045", "matrix": [2, 5], "x": 5.25, "y": 2, "w": 1.5 },
+                { "label":"K046", "matrix": [2, 6], "x": 6.75, "y": 2, "w": 1 },
+                { "label":"K047", "matrix": [2, 7], "x": 7.75, "y": 2, "w": 1 },
+                { "label":"K048", "matrix": [2, 8], "x": 8.75, "y": 2, "w": 1 },
+                { "label":"K049", "matrix": [2, 9], "x": 9.75, "y": 2, "w": 1 },
+                { "label":"K050", "matrix": [7, 5], "x": 10.75, "y": 2, "w": 1 },
+                { "label":"K051", "matrix": [7, 6], "x": 11.75, "y": 2, "w": 1 },
+                { "label":"K052", "matrix": [7, 7], "x": 12.75, "y": 2, "w": 1 },
+                { "label":"K053", "matrix": [7, 8], "x": 13.75, "y": 2, "w": 1 },
+                { "label":"K054", "matrix": [7, 9], "x": 14.75, "y": 2, "w": 1 },
+                { "label":"K055", "matrix": [9, 0], "x": 15.75, "y": 2, "w": 1 },
+                { "label":"K056", "matrix": [9, 1], "x": 16.75, "y": 2, "w": 1 },
+                { "label":"K057", "matrix": [9, 2], "x": 17.75, "y": 2, "w": 1 },
+                { "label":"K058", "matrix": [9, 3], "x": 18.75, "y": 2, "w": 1.5 },
```
Something's not quite right — looks like you need to adjust the `y` coordinates for many keys by `0.25`:
![Layout as shown on config.qmk.fm](https://github.com/qmk/qmk_firmware/assets/616082/ccc9a007-a6b7-4a0b-83a8-c5b86c37837b)
(To see the representation of your layout, you need to open https://config.qmk.fm, click once near the keyboard layout that appears to set the input focus properly, then press Ctrl+Shift+I and upload your `info.json` file.  Also switch legends to “Index” in the settings to confirm that the key ordering is correct; you may also use the “Matrix” setting to check the matrix row,col numbers.)

TokageItLab replied:
Oh, you are right, thanks.

sigprof commented on keyboards/varanidae/readme.md:
```diff
@@ -0,0 +1,24 @@
+
+Enter the bootloader in 3 ways:
+
+* **Bootmagic reset**: Hold down the key at (0,0) in the matrix (usually the top left key or Escape) and plug in the keyboard
```
You should write something more precise here — your (0,0) key is certainly not Esc (it appears to be F16 in the default keymap, and the top left location is occupied by an encoder).

Alternatively, you can redefine the bootmagic key to use another matrix location (including the key which you intend to be Esc).

TokageItLab replied:
Replaced "Esc" with "F16 as default".

TokageItLab replied:
Finally, the 0,0 element of the matrix became an encoder push switch so I changed the description to that.

sigprof commented on keyboards/varanidae/info.json:
```diff
@@ -0,0 +1,155 @@
+        "bootmagic": true,
+        "extrakey": true,
+        "mousekey": true,
+        "dip_switch": true,
```
```suggestion
```

sigprof commented on keyboards/varanidae/info.json:
```diff
@@ -0,0 +1,155 @@
+    },
+    "diode_direction": "COL2ROW",
+    "matrix_pins": {
+        "cols": ["F4", "D4", "D7", "B4", "B7", "C7", "F0", "B2", "B1", "F6"],
```
```suggestion
        "cols": ["F4", "D4", "D7", "B4", "B7", "C7", "F0", "B2", "B1", "F6", "D2"],
```

sigprof commented on keyboards/varanidae/info.json:
```diff
@@ -0,0 +1,155 @@
+    "layouts": {
+        "LAYOUT": {
+            "layout": [
+                { "label":"K001", "matrix": [0, 0], "x": 1, "y": 0, "w": 1 },
```
```suggestion
                { "label":"e0", "matrix": [0, 10], "x": 0, "y": 0, "w": 1 },
                { "label":"K001", "matrix": [0, 0], "x": 1, "y": 0, "w": 1 },
```

sigprof commented on keyboards/varanidae/keymaps/default/keymap.c:
```diff
@@ -0,0 +1,35 @@
+
+const uint16_t PROGMEM keymaps[][MATRIX_ROWS][MATRIX_COLS] = {
+    [0] = LAYOUT(
+                 KC_F16, KC_F17,  KC_F18,  KC_F19,    KC_F1,     KC_F2, KC_F3, KC_F4, KC_F5, KC_F6, KC_F7, KC_F8, KC_F9,   KC_F10, KC_F11, KC_F12,  KC_F13, KC_F14, KC_F15,
+        KC_F20,  KC_F21, KC_F22,  KC_F23,  KC_F24,    KC_ESC,    KC_1,  KC_2,  KC_3,  KC_4,  KC_5,  KC_6,  KC_7,  KC_8,    KC_9,   KC_0,   KC_MINS, KC_EQL, KC_DEL, KC_BSPC,
```
```suggestion
        KC_BTN3, KC_F16, KC_F17,  KC_F18,  KC_F19,    KC_F1,     KC_F2, KC_F3, KC_F4, KC_F5, KC_F6, KC_F7, KC_F8, KC_F9,   KC_F10, KC_F11, KC_F12,  KC_F13, KC_F14, KC_F15,
        KC_F20,  KC_F21, KC_F22,  KC_F23,  KC_F24,    KC_ESC,    KC_1,  KC_2,  KC_3,  KC_4,  KC_5,  KC_6,  KC_7,  KC_8,    KC_9,   KC_0,   KC_MINS, KC_EQL, KC_DEL, KC_BSPC,
```

sigprof commented on keyboards/varanidae/keymaps/via/keymap.c:
```diff
@@ -0,0 +1,35 @@
+
+const uint16_t PROGMEM keymaps[][MATRIX_ROWS][MATRIX_COLS] = {
+    [0] = LAYOUT(
+                 KC_F16, KC_F17,  KC_F18,  KC_F19,    KC_F1,     KC_F2, KC_F3, KC_F4, KC_F5, KC_F6, KC_F7, KC_F8, KC_F9,   KC_F10, KC_F11, KC_F12,  KC_F13, KC_F14, KC_F15,
+        KC_F20,  KC_F21, KC_F22,  KC_F23,  KC_F24,    KC_ESC,    KC_1,  KC_2,  KC_3,  KC_4,  KC_5,  KC_6,  KC_7,  KC_8,    KC_9,   KC_0,   KC_MINS, KC_EQL, KC_DEL, KC_BSPC,
```
```suggestion
        KC_BTN3, KC_F16, KC_F17,  KC_F18,  KC_F19,    KC_F1,     KC_F2, KC_F3, KC_F4, KC_F5, KC_F6, KC_F7, KC_F8, KC_F9,   KC_F10, KC_F11, KC_F12,  KC_F13, KC_F14, KC_F15,
        KC_F20,  KC_F21, KC_F22,  KC_F23,  KC_F24,    KC_ESC,    KC_1,  KC_2,  KC_3,  KC_4,  KC_5,  KC_6,  KC_7,  KC_8,    KC_9,   KC_0,   KC_MINS, KC_EQL, KC_DEL, KC_BSPC,
```

sigprof reviewed (commented):
Does this set of changes work on the real hardware?  If it does, you can drop the DIP switch code, and then add the `MATRIX_MASKED` optimization.

Note that the VIA JSON would need to be updated accordingly to account for the new matrix dimensions.

TokageItLab commented:
@sigprof  I already did that same thing, but it determined that multiple unrelated keys were being pressed at the same time, which messed up the signal.

I thought that I could add another row to row as well as col to make better use of MASK like:

0b111110
0b111110
0b111110
0b000001 <- RE

but there is no more free pin.

sigprof commented:
> 0b111110
> 0b111110
> 0b111110
> 0b000001 <- RE

An extra row should not be needed; the mask should look like
```
0b11111111111
0b01111111111
0b01111111111
```
(the LSB corresponds to column 0, so the bit mask looks reversed)

TokageItLab commented:
```
0b11111111111
0b01111111111
0b01111111111
```

@sigprof I've already tried that same thing, but it didn't change the results.

TokageItLab commented:
> but it didn't change the results.

@sigprof To say that the result does not change is precisely incorrect: somehow the scrolling also stops responding at the same time as the encoder push, and the signals for the unrelated keys are messed up in a different pattern.

TokageItLab commented:
@sigprof Encoder push problem has been resolved.

As a result, it is a electric signal order problem with circulating flow, so if you adding an independent switch to the first index instead of the last index, seems to be correctly detected signal. It was a hassle to shift the matrix for all the keys, but I've done it and I believe now this is fine.

Thanks a lot for your advices especially about the EEPROM was helpful.

sigprof commented on keyboards/varanidae/info.json:
```diff
@@ -0,0 +1,156 @@
+        "rows": ["B0", "F5", "C6", "E6", "B5", "D5", "F1", "B6", "B3", "F7", "D3"]
+    },
+    "layouts": {
+        "LAYOUT": {
```
Apparently your keyboard supports multiple possible layouts (there are options for split left Shift, split right Shift, split Backspace, ISO Enter).  The way to support such keyboards in QMK is:

1. Create a layout definition which covers all possible matrix positions, named `LAYOUT_all`.
   
   The current `LAYOUT` definition already does that, so the only thing needed is to rename it to `LAYOUT_all`.

2. Create some layout definitions for commonly used layouts (covering all possible permutations is not required, but at least the common ANSI and ISO layout options should be covered).

   For this keyboard probably at least definitions like `LAYOUT_ansi` and `LAYOUT_iso` should be supplied (just copy the `LAYOUT_all` definition and adjust the coordinates and sizes of affected keys, removing the keys which are not present in the specific layout):
   - `LAYOUT_ansi` — ANSI Enter, 2u Backspace, 2.25u left Shift, 2.75u right Shift.
   - `LAYOUT_iso` — ISO Enter, 2u Backspace, split (1.75u+1u) left Shift, 2.75u right Shift.

   Although given that split right Shift is advertised as an important feature of your keyboard, you may decide to supply something like `LAYOUT_ansi_split_bs_rshift` and `LAYOUT_iso_split_bs_rshift` too.

   It is not required to supply specific keymaps for all those layouts (your keyboard is wired so that the default keymap should give reasonable results for non-split Shift/Backspace and ISO Enter); you may decide to add some just to test them though.

Note that for ISO layouts the ISO Enter key should be placed in the layout at the bottom of two rows it occupies (so that it ends up on the same row as ANSI Enter would); the key ordering would be like this:
![ISO Enter key ordering](https://github.com/qmk/qmk_firmware/assets/616082/541c021a-99d7-409e-8848-a06b065b1692)


```suggestion
        "LAYOUT_all": {
```

sigprof commented on keyboards/varanidae/keymaps/default/keymap.c:
```diff
@@ -0,0 +1,35 @@
+#include QMK_KEYBOARD_H
+
+const uint16_t PROGMEM keymaps[][MATRIX_ROWS][MATRIX_COLS] = {
+    [0] = LAYOUT(
```
```suggestion
    [0] = LAYOUT_all(
```

sigprof commented on keyboards/varanidae/keymaps/via/keymap.c:
```diff
@@ -0,0 +1,35 @@
+#include QMK_KEYBOARD_H
+
+const uint16_t PROGMEM keymaps[][MATRIX_ROWS][MATRIX_COLS] = {
+    [0] = LAYOUT(
```
```suggestion
    [0] = LAYOUT_all(
```

sigprof commented on keyboards/varanidae/keymaps/default/keymap.c:
```diff
@@ -0,0 +1,35 @@
+    [0] = LAYOUT(
+        KC_BTN3,    KC_F16, KC_F17,  KC_F18,  KC_F19,    KC_F1,     KC_F2, KC_F3, KC_F4, KC_F5, KC_F6, KC_F7, KC_F8, KC_F9,   KC_F10, KC_F11, KC_F12,  KC_F13, KC_F14, KC_F15,
+        KC_F20,     KC_F21, KC_F22,  KC_F23,  KC_F24,    KC_ESC,    KC_1,  KC_2,  KC_3,  KC_4,  KC_5,  KC_6,  KC_7,  KC_8,    KC_9,   KC_0,   KC_MINS, KC_EQL, KC_DEL, KC_BSPC,
+        KC_PSLS,    KC_P7,  KC_P8,   KC_P9,   KC_HOME,   KC_TAB,      KC_Q,  KC_W,  KC_E,  KC_R,  KC_T,  KC_Y,  KC_U,  KC_I,    KC_O,   KC_P,   KC_LBRC,  KC_RBRC,  KC_ENT,
+        KC_PAST,    KC_P4,  KC_P5,   KC_P6,   KC_END,    KC_LALT,        KC_A,  KC_S,  KC_D,  KC_F,  KC_G,  KC_H,  KC_J,  KC_K,    KC_L,   KC_SCLN, KC_QUOT,  KC_BSLS,
```
The assignments of `KC_BSLS` and `KC_ENT` look wrong — for the ANSI layout the backslash key is above the Enter key, not below it.  If the wrong-looking code actually produced the expected mapping for the backslash and Enter keys, probably the layout definition needs to be corrected to match the actual PCB wiring in this area.
```suggestion
        KC_PSLS,    KC_P7,  KC_P8,   KC_P9,   KC_HOME,   KC_TAB,      KC_Q,  KC_W,  KC_E,  KC_R,  KC_T,  KC_Y,  KC_U,  KC_I,    KC_O,   KC_P,   KC_LBRC,  KC_RBRC,  KC_BSLS,
        KC_PAST,    KC_P4,  KC_P5,   KC_P6,   KC_END,    KC_LALT,        KC_A,  KC_S,  KC_D,  KC_F,  KC_G,  KC_H,  KC_J,  KC_K,    KC_L,   KC_SCLN, KC_QUOT,  KC_ENT,
```

TokageItLab replied:
I'm thinking of ISO Enter as the default layout for this keyboard, so KC_ENT is on top. are we forced to default to ANSI Enter?

TokageItLab replied:
Ah, I understand. The problem is that K058 in LAYOUT is not in ISO shape; I specify "w": 1.25, "h": 2 for K058 to make it consistent. Is this correct?

TokageItLab replied:
For now, I changed this to:
```
KC_LBRC,  KC_RBRC, KC_BSLS,
KC_SCLN, KC_QUOT,  KC_ENT,
```
for ansi,

and
```
KC_LBRC,  KC_RBRC,
KC_QUOT,  KC_NUBS, KC_ENT,
```

for iso and default.

sigprof commented on keyboards/varanidae/keymaps/default/keymap.c:
```diff
@@ -0,0 +1,35 @@
+        KC_F20,     KC_F21, KC_F22,  KC_F23,  KC_F24,    KC_ESC,    KC_1,  KC_2,  KC_3,  KC_4,  KC_5,  KC_6,  KC_7,  KC_8,    KC_9,   KC_0,   KC_MINS, KC_EQL, KC_DEL, KC_BSPC,
+        KC_PSLS,    KC_P7,  KC_P8,   KC_P9,   KC_HOME,   KC_TAB,      KC_Q,  KC_W,  KC_E,  KC_R,  KC_T,  KC_Y,  KC_U,  KC_I,    KC_O,   KC_P,   KC_LBRC,  KC_RBRC,  KC_ENT,
+        KC_PAST,    KC_P4,  KC_P5,   KC_P6,   KC_END,    KC_LALT,        KC_A,  KC_S,  KC_D,  KC_F,  KC_G,  KC_H,  KC_J,  KC_K,    KC_L,   KC_SCLN, KC_QUOT,  KC_BSLS,
+        KC_PMNS,    KC_P1,  KC_P2,   KC_P3,   KC_PGUP,   KC_LSFT, KC_BSLS,  KC_Z,  KC_X,  KC_C,  KC_V,  KC_B,  KC_N,  KC_M,  KC_COMM, KC_DOT, KC_SLSH, KC_GRV,   KC_RSFT,
```
The proper mapping for the extra ISO key near the left Shift is `KC_NUBS`.

(Technically the key near the ISO Enter should also be `KC_NUHS` instead of `KC_BSLS` for an ISO keyboard, but most OSes treat those keycodes as mostly equivalent, so using the ANSI mapping for that key is acceptable; the key near the left Shift really must be NUBS though.)
```suggestion
        KC_PMNS,    KC_P1,  KC_P2,   KC_P3,   KC_PGUP,   KC_LSFT, KC_NUBS,  KC_Z,  KC_X,  KC_C,  KC_V,  KC_B,  KC_N,  KC_M,  KC_COMM, KC_DOT, KC_SLSH, KC_GRV,   KC_RSFT,
```

sigprof commented on keyboards/varanidae/keymaps/via/keymap.c:
```diff
@@ -0,0 +1,35 @@
+    [0] = LAYOUT(
+        KC_BTN3,    KC_F16, KC_F17,  KC_F18,  KC_F19,    KC_F1,     KC_F2, KC_F3, KC_F4, KC_F5, KC_F6, KC_F7, KC_F8, KC_F9,   KC_F10, KC_F11, KC_F12,  KC_F13, KC_F14, KC_F15,
+        KC_F20,     KC_F21, KC_F22,  KC_F23,  KC_F24,    KC_ESC,    KC_1,  KC_2,  KC_3,  KC_4,  KC_5,  KC_6,  KC_7,  KC_8,    KC_9,   KC_0,   KC_MINS, KC_EQL, KC_DEL, KC_BSPC,
+        KC_PSLS,    KC_P7,  KC_P8,   KC_P9,   KC_HOME,   KC_TAB,      KC_Q,  KC_W,  KC_E,  KC_R,  KC_T,  KC_Y,  KC_U,  KC_I,    KC_O,   KC_P,   KC_LBRC,  KC_RBRC,  KC_ENT,
+        KC_PAST,    KC_P4,  KC_P5,   KC_P6,   KC_END,    KC_LALT,        KC_A,  KC_S,  KC_D,  KC_F,  KC_G,  KC_H,  KC_J,  KC_K,    KC_L,   KC_SCLN, KC_QUOT,  KC_BSLS,
```
```suggestion
        KC_PSLS,    KC_P7,  KC_P8,   KC_P9,   KC_HOME,   KC_TAB,      KC_Q,  KC_W,  KC_E,  KC_R,  KC_T,  KC_Y,  KC_U,  KC_I,    KC_O,   KC_P,   KC_LBRC,  KC_RBRC,  KC_BSLS,
        KC_PAST,    KC_P4,  KC_P5,   KC_P6,   KC_END,    KC_LALT,        KC_A,  KC_S,  KC_D,  KC_F,  KC_G,  KC_H,  KC_J,  KC_K,    KC_L,   KC_SCLN, KC_QUOT,  KC_ENT,
```

TokageItLab replied:
As described in https://github.com/qmk/qmk_firmware/pull/21340#discussion_r1241316148, perhaps this issue is a wrong shape of LAYOUT and not a wrong matrix.

sigprof commented on keyboards/varanidae/keymaps/via/keymap.c:
```diff
@@ -0,0 +1,35 @@
+        KC_F20,     KC_F21, KC_F22,  KC_F23,  KC_F24,    KC_ESC,    KC_1,  KC_2,  KC_3,  KC_4,  KC_5,  KC_6,  KC_7,  KC_8,    KC_9,   KC_0,   KC_MINS, KC_EQL, KC_DEL, KC_BSPC,
+        KC_PSLS,    KC_P7,  KC_P8,   KC_P9,   KC_HOME,   KC_TAB,      KC_Q,  KC_W,  KC_E,  KC_R,  KC_T,  KC_Y,  KC_U,  KC_I,    KC_O,   KC_P,   KC_LBRC,  KC_RBRC,  KC_ENT,
+        KC_PAST,    KC_P4,  KC_P5,   KC_P6,   KC_END,    KC_LALT,        KC_A,  KC_S,  KC_D,  KC_F,  KC_G,  KC_H,  KC_J,  KC_K,    KC_L,   KC_SCLN, KC_QUOT,  KC_BSLS,
+        KC_PMNS,    KC_P1,  KC_P2,   KC_P3,   KC_PGUP,   KC_LSFT, KC_BSLS,  KC_Z,  KC_X,  KC_C,  KC_V,  KC_B,  KC_N,  KC_M,  KC_COMM, KC_DOT, KC_SLSH, KC_GRV,   KC_RSFT,
```
```suggestion
        KC_PMNS,    KC_P1,  KC_P2,   KC_P3,   KC_PGUP,   KC_LSFT, KC_NUBS,  KC_Z,  KC_X,  KC_C,  KC_V,  KC_B,  KC_N,  KC_M,  KC_COMM, KC_DOT, KC_SLSH, KC_GRV,   KC_RSFT,
```

sigprof commented on keyboards/varanidae/info.json:
```diff
@@ -0,0 +1,156 @@
+                { "label":"K055", "matrix": [9, 1], "x": 15.75, "y": 2.25, "w": 1 },
+                { "label":"K056", "matrix": [9, 2], "x": 16.75, "y": 2.25, "w": 1 },
+                { "label":"K057", "matrix": [9, 3], "x": 17.75, "y": 2.25, "w": 1 },
+                { "label":"K058", "matrix": [9, 4], "x": 18.75, "y": 2.25, "w": 1.5 },
```
I looked at the PCB, and apparently K058 is ANSI/ISO Enter, while K076 is ANSI backslash/ISO NUHS; the layout definition uses ANSI-shaped keys, so these key positions need to be swapped accordingly.
```suggestion
                { "label":"K076", "matrix": [10, 3], "x": 18.75, "y": 2.25, "w": 1.5 },
```

TokageItLab replied:
As described in https://github.com/qmk/qmk_firmware/pull/21340#discussion_r1241316148, perhaps this issue is a wrong shape of LAYOUT and not a wrong matrix.

sigprof commented on keyboards/varanidae/info.json:
```diff
@@ -0,0 +1,156 @@
+                { "label":"K073", "matrix": [8, 10], "x": 15, "y": 3.25, "w": 1 },
+                { "label":"K074", "matrix": [10, 1], "x": 16, "y": 3.25, "w": 1 },
+                { "label":"K075", "matrix": [10, 2], "x": 17, "y": 3.25, "w": 1 },
+                { "label":"K076", "matrix": [10, 3], "x": 18, "y": 3.25, "w": 2.25 },
```
```suggestion
                { "label":"K058", "matrix": [9, 4], "x": 18, "y": 3.25, "w": 2.25 },
```

TokageItLab replied:
As described in https://github.com/qmk/qmk_firmware/pull/21340#discussion_r1241316148, perhaps this issue is a wrong shape of LAYOUT and not a wrong matrix.

sigprof reviewed (changes requested):
Some adjustment for proper support of multiple layouts are needed (the [PR Checklist](https://docs.qmk.fm/#/pr_checklist) requires that for any keyboards that support multiple layouts; just implementing what amounts to `LAYOUT_all` and having the proper definitions for multiple layouts only in VIA is frowned upon).

And apparently there is some confusion with the Enter and backslash keys, which needs to be fixed.

TokageItLab commented:
I fixed some indexes in layout and added keymaps for `ansi`, `iso`, `ansi_split_bs_rshift`, and `iso_split_bs_rshift`. Also, the labels have been made so that the actual keys are easily recognizable rather than sequential numbers.

> Note that for ISO layouts the ISO Enter key should be placed in the layout at the bottom of two rows it occupies (so that it ends up on the same row as ANSI Enter would); the key ordering would be like this:

@sigprof If I refer to the picture, is this correct for iso layout? `KC_QUOT,  KC_BSLS, KC_ENT,`

sigprof commented:
> If I refer to the picture, is this correct for iso layout? `KC_QUOT, KC_BSLS, KC_ENT,`

Almost; you may look at the various ISO layouts under `layouts/default` — they have `KC_QUOT, KC_NUHS, KC_ENT,`.  Using `KC_NUHS` should be better if you are explicitly making the ISO layout.

sigprof commented:
Also you may replace the full `LAYOUT_all` definition with just
```json
    "layout_aliases": {
        "LAYOUT_all": "LAYOUT_iso_split_bs_rshift"
    },
```
(I was writing some suggestions, but apparently you are actively changing the code now…)

TokageItLab commented:
@sigprof 

> Also you may replace the full LAYOUT_all definition with just

Thanks, I done that.

Sorry, I changed all KC_BSLS to KC_NUBS in iso layout. Now I finished the fix.

TokageItLab commented:
> Almost; you may look at the various ISO layouts under layouts/default — they have KC_QUOT, KC_NUHS, KC_ENT,. Using KC_NUHS should be better if you are explicitly making the ISO layout.

Oh, I thought that was a typo because I didn't want to put the '#' there, but in most environments it would be '\'. I did not know that.

Thanks, I fixed it. I think that's really the end of it.

sigprof commented on keyboards/varanidae/info.json:
```diff
@@ -0,0 +1,515 @@
+                { "label":";", "matrix": [10, 1], "x": 16, "y": 3.25, "w": 1 },
+                { "label":"'", "matrix": [10, 2], "x": 17, "y": 3.25, "w": 1 },
+                { "label":"\\", "matrix": [10, 3], "x": 18, "y": 3.25, "w": 1 },
+                { "label":"Enter", "matrix": [9, 4], "x": 18.75, "y": 2.25, "w": 1.25, "h": 2 },
```
ISO Enter needs to be positioned as if it was a 1.25×2 rectangle (the extra 0.25u extension at the left side is added implicitly by the GUI):
![ISO Enter](https://github.com/qmk/qmk_firmware/assets/616082/fa2093bc-8e3d-4c14-9f8b-d4440a1ff29a)
```suggestion
                { "label":"Enter", "matrix": [9, 4], "x": 19, "y": 2.25, "w": 1.25, "h": 2 },
```

sigprof commented on keyboards/varanidae/info.json:
```diff
@@ -0,0 +1,515 @@
+                { "label":";", "matrix": [10, 1], "x": 16, "y": 3.25, "w": 1 },
+                { "label":"'", "matrix": [10, 2], "x": 17, "y": 3.25, "w": 1 },
+                { "label":"\\", "matrix": [10, 3], "x": 18, "y": 3.25, "w": 1 },
+                { "label":"Enter", "matrix": [9, 4], "x": 18.75, "y": 2.25, "w": 1.25, "h": 2 },
```
```suggestion
                { "label":"Enter", "matrix": [9, 4], "x": 19, "y": 2.25, "w": 1.25, "h": 2 },
```

sigprof commented on keyboards/varanidae/keymaps/ansi/keymap.c:
```diff
@@ -0,0 +1,35 @@
+        KC_F20,     KC_F21, KC_F22,  KC_F23,  KC_F24,    KC_ESC,    KC_1,  KC_2,  KC_3,  KC_4,  KC_5,  KC_6,  KC_7,  KC_8,    KC_9,   KC_0,   KC_MINS, KC_EQL, KC_BSPC,
+        KC_PSLS,    KC_P7,  KC_P8,   KC_P9,   KC_HOME,   KC_TAB,      KC_Q,  KC_W,  KC_E,  KC_R,  KC_T,  KC_Y,  KC_U,  KC_I,    KC_O,   KC_P,   KC_LBRC,  KC_RBRC, KC_BSLS,
+        KC_PAST,    KC_P4,  KC_P5,   KC_P6,   KC_END,    KC_LALT,        KC_A,  KC_S,  KC_D,  KC_F,  KC_G,  KC_H,  KC_J,  KC_K,    KC_L,   KC_SCLN, KC_QUOT,  KC_ENT,
+        KC_PMNS,    KC_P1,  KC_P2,   KC_P3,   KC_PGUP,   KC_LSFT,  KC_Z,  KC_X,  KC_C,  KC_V,  KC_B,  KC_N,  KC_M,  KC_COMM, KC_DOT, KC_SLSH, KC_RSFT,
```
```suggestion
        KC_PMNS,    KC_P1,  KC_P2,   KC_P3,   KC_PGUP,   KC_LSFT,           KC_Z,  KC_X,  KC_C,  KC_V,  KC_B,  KC_N,  KC_M,  KC_COMM, KC_DOT, KC_SLSH, KC_RSFT,
```

sigprof reviewed (changes requested):
There is an issue with the ISO enter positioning, and a very minor formatting inconsistency, but apart from that, everything else seems fine.

TokageItLab commented:
@sigprof Fixed, thank you for a lot of checking. These firmware configurations is unknown territory for me, so that is very helpful. Thanks again.

TokageItLab commented:
@drashna I think all the fixes are done, is there anything else that needs to be worked on?
